collections/accounts.go

```diff
@@ -0,0 +1,526 @@
+package collections
+
+import (
+	"context"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/gosimple/slug"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/textileio/go-threads/core/thread"
+	"github.com/textileio/textile/util"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var (
+	usernameRx *regexp.Regexp
+
+	ErrInvalidUsername = fmt.Errorf("username may only contain alphanumeric characters or single hyphens, and cannot begin or end with a hyphen")
+)
+
+func init() {
+	usernameRx = regexp.MustCompile(`^[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)?$`)
+}
+
+type Account struct {
+	Type      AccountType
+	Key       crypto.PubKey
+	Secret    crypto.PrivKey
+	Name      string
+	Username  string
+	Email     string
+	Token     thread.Token
+	Members   []Member
+	CreatedAt time.Time
+}
+
+type AccountType int
+
+const (
+	Dev AccountType = iota
+	Org
+)
+
+type Member struct {
+	Key      crypto.PubKey
+	Username string
+	Role     Role
+}
+
+type Role int
+
+const (
+	OrgOwner Role = iota
+	OrgMember
+)
+
+func (r Role) String() (s string) {
+	switch r {
+	case OrgOwner:
+		s = "owner"
+	case OrgMember:
+		s = "member"
+	}
+	return
+}
+
+func NewDevContext(ctx context.Context, dev *Account) context.Context {
+	return context.WithValue(ctx, ctxKey("developer"), dev)
+}
+
+func DevFromContext(ctx context.Context) (*Account, bool) {
+	dev, ok := ctx.Value(ctxKey("developer")).(*Account)
+	return dev, ok
+}
+
+func NewOrgContext(ctx context.Context, org *Account) context.Context {
+	return context.WithValue(ctx, ctxKey("org"), org)
+}
+
+func OrgFromContext(ctx context.Context) (*Account, bool) {
+	org, ok := ctx.Value(ctxKey("org")).(*Account)
+	return org, ok
+}
+
+type Accounts struct {
+	col *mongo.Collection
+}
+
+func NewAccounts(ctx context.Context, db *mongo.Database) (*Accounts, error) {
+	t := &Accounts{col: db.Collection("accounts")}
+	_, err := t.col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys: bson.D{{"username", 1}},
+			Options: options.Index().SetUnique(true).
+				SetCollation(&options.Collation{Locale: "en", Strength: 2}),
```

Comment:
Ensures that usernames are unique across cases (upper and lower). For example, `mike` is the same as `Mike`.